<center><b><big> CITS 5508 LAB SHEET 3</big></b></center>

**Name: Thanh Duy Tang**<br>
**Student Number: 22803018**<br>
**Date created: 24th March 2020 <br>
**Last modified: 17th Apr 2020

## 1. Setup

Before going to loading data part, we need to make sure that MatplotLib figures inline and prepare a function to save the figures. Moreover, we should use Python 3.0 as well as Scikit-Learn >= 0.20

In [181]:
#Python >= 3.0 is required
import sys
assert sys.version_info >= (3,5)

#Scikit-Learn >= 0.20 
import sklearn
assert sklearn.__version__ >= '0.20'

#Get the pretty figures
%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
mpl.rc('axes', labelsize=14)
mpl.rc('xtick', labelsize=12)
mpl.rc('ytick', labelsize=12)

#Save the figures
import os
LAB2_ROOT_DIR = "."
CHAPTER_ID = "LAB SHEET 2"
IMAGES_PATH = os.path.join(LAB2_ROOT_DIR, "images", CHAPTER_ID)
os.makedirs(IMAGES_PATH, exist_ok=True)

def save_fig(fig_id, tight_layout=True, fig_extension="png", resolution=300):
    path = os.path.join(IMAGES_PATH, fig_id + "." + fig_extension)
    print("Saving figure", fig_id)
    if tight_layout:
        plt.tight_layout()
    plt.savefig(path, format=fig_extension, dpi=resolution)
    
#Ignore useless warnings
import warnings
warnings.filterwarnings(action="ignore", message="^internal gelsd")

## 2. Loading data
First, we need to dowload the zip file from http://archive.ics.uci.edu/ml/datasets/Parkinsons
and put the file into the same directory of this lab file.

In [182]:
#Load the Pandas libraries with alias "pd"
import os
import pandas as pd
#Read data from "parkinsons.data"
data_set = pd.read_csv("parkinsons.data", delimiter = ',')
data_set

name  MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
0    phon_R01_S01_1      119.992       157.302        74.997         0.00784   
1    phon_R01_S01_2      122.400       148.650       113.819         0.00968   
2    phon_R01_S01_3      116.682       131.111       111.555         0.01050   
3    phon_R01_S01_4      116.676       137.871       111.366         0.00997   
4    phon_R01_S01_5      116.014       141.781       110.655         0.01284   
..              ...          ...           ...           ...             ...   
190  phon_R01_S50_2      174.188       230.978        94.261         0.00459   
191  phon_R01_S50_3      209.516       253.017        89.488         0.00564   
192  phon_R01_S50_4      174.688       240.005        74.287         0.01360   
193  phon_R01_S50_5      198.764       396.961        74.904         0.00740   
194  phon_R01_S50_6      214.289       260.277        77.973         0.00567   

     MDVP:Jitter(Abs)  MDVP:RAP  MDVP:PPQ  Jitter:DDP  MDVP:Shimmer  ...  \
0             0.00007   0.00370   0.00554     0.01109       0.04374  ...   
1             0.00008   0.00465   0.00696     0.01394       0.06134  ...   
2             0.00009   0.00544   0.00781     0.01633       0.05233  ...   
3             0.00009   0.00502   0.00698     0.01505       0.05492  ...   
4             0.00011   0.00655   0.00908     0.01966       0.06425  ...   
..                ...       ...       ...         ...           ...  ...   
190           0.00003   0.00263   0.00259     0.00790       0.04087  ...   
191           0.00003   0.00331   0.00292     0.00994       0.02751  ...   
192           0.00008   0.00624   0.00564     0.01873       0.02308  ...   
193           0.00004   0.00370   0.00390     0.01109       0.02296  ...   
194           0.00003   0.00295   0.00317     0.00885       0.01884  ...   

     Shimmer:DDA      NHR     HNR  status      RPDE       DFA   spread1  \
0        0.06545  0.02211  21.033       1  0.414783  0.815285 -4.813031   
1        0.09403  0.01929  19.085       1  0.458359  0.819521 -4.075192   
2        0.08270  0.01309  20.651       1  0.429895  0.825288 -4.443179   
3        0.08771  0.01353  20.644       1  0.434969  0.819235 -4.117501   
4        0.10470  0.01767  19.649       1  0.417356  0.823484 -3.747787   
..           ...      ...     ...     ...       ...       ...       ...   
190      0.07008  0.02764  19.517       0  0.448439  0.657899 -6.538586   
191      0.04812  0.01810  19.147       0  0.431674  0.683244 -6.195325   
192      0.03804  0.10715  17.883       0  0.407567  0.655683 -6.787197   
193      0.03794  0.07223  19.020       0  0.451221  0.643956 -6.744577   
194      0.03078  0.04398  21.209       0  0.462803  0.664357 -5.724056   

      spread2        D2       PPE  
0    0.266482  2.301442  0.284654  
1    0.335590  2.486855  0.368674  
2    0.311173  2.342259  0.332634  
3    0.334147  2.405554  0.368975  
4    0.234513  2.332180  0.410335  
..        ...       ...       ...  
190  0.121952  2.657476  0.133050  
191  0.129303  2.784312  0.168895  
192  0.158453  2.679772  0.131728  
193  0.207454  2.138608  0.123306  
194  0.190667  2.555477  0.148569  

[195 rows x 24 columns]

In [183]:
data_set.describe()

MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
count   195.000000    195.000000    195.000000      195.000000   
mean    154.228641    197.104918    116.324631        0.006220   
std      41.390065     91.491548     43.521413        0.004848   
min      88.333000    102.145000     65.476000        0.001680   
25%     117.572000    134.862500     84.291000        0.003460   
50%     148.790000    175.829000    104.315000        0.004940   
75%     182.769000    224.205500    140.018500        0.007365   
max     260.105000    592.030000    239.170000        0.033160   

       MDVP:Jitter(Abs)    MDVP:RAP    MDVP:PPQ  Jitter:DDP  MDVP:Shimmer  \
count        195.000000  195.000000  195.000000  195.000000    195.000000   
mean           0.000044    0.003306    0.003446    0.009920      0.029709   
std            0.000035    0.002968    0.002759    0.008903      0.018857   
min            0.000007    0.000680    0.000920    0.002040      0.009540   
25%            0.000020    0.001660    0.001860    0.004985      0.016505   
50%            0.000030    0.002500    0.002690    0.007490      0.022970   
75%            0.000060    0.003835    0.003955    0.011505      0.037885   
max            0.000260    0.021440    0.019580    0.064330      0.119080   

       MDVP:Shimmer(dB)  ...  Shimmer:DDA         NHR         HNR      status  \
count        195.000000  ...   195.000000  195.000000  195.000000  195.000000   
mean           0.282251  ...     0.046993    0.024847   21.885974    0.753846   
std            0.194877  ...     0.030459    0.040418    4.425764    0.431878   
min            0.085000  ...     0.013640    0.000650    8.441000    0.000000   
25%            0.148500  ...     0.024735    0.005925   19.198000    1.000000   
50%            0.221000  ...     0.038360    0.011660   22.085000    1.000000   
75%            0.350000  ...     0.060795    0.025640   25.075500    1.000000   
max            1.302000  ...     0.169420    0.314820   33.047000    1.000000   

             RPDE         DFA     spread1     spread2          D2         PPE  
count  195.000000  195.000000  195.000000  195.000000  195.000000  195.000000  
mean     0.498536    0.718099   -5.684397    0.226510    2.381826    0.206552  
std      0.103942    0.055336    1.090208    0.083406    0.382799    0.090119  
min      0.256570    0.574282   -7.964984    0.006274    1.423287    0.044539  
25%      0.421306    0.674758   -6.450096    0.174351    2.099125    0.137451  
50%      0.495954    0.722254   -5.720868    0.218885    2.361532    0.194052  
75%      0.587562    0.761881   -5.046192    0.279234    2.636456    0.252980  
max      0.685151    0.825288   -2.434031    0.450493    3.671155    0.527367  

[8 rows x 23 columns]

In [184]:
#Count the value
data_set['name'].value_counts()

phon_R01_S10_1    1
phon_R01_S50_2    1
phon_R01_S02_3    1
phon_R01_S25_4    1
phon_R01_S04_2    1
                 ..
phon_R01_S44_3    1
phon_R01_S06_1    1
phon_R01_S05_2    1
phon_R01_S17_3    1
phon_R01_S13_2    1
Name: name, Length: 195, dtype: int64

As we can see, the column 'name' has distinct values, so we can easily set that column as index column.

In [185]:
#Set the "name" as index column because this column has distinct values.
data_set.set_index('name', inplace = True)

## 3. Cleaning and preparing data

We don't know whether delete column or not delete column can get better result. So I decided to do both, which means that I keep the columns and delete columns, and then compare each other. When we have a look at the describe of the dataset, I decided to delete some columns whose means are relatively small. Those columns are MDVP:Jitter(%), MDVP:Jitter(Abs), MDVP:RAP, MDVP:PPQ, Jitter:DDP, MDVP:Shimmer, Shimmer:APQ3, Shimmer:APQ5, DVP:APQ, Shimmer:DDA, NHR. I also changed the value of column 'status' with 0(healthy) and 1(unhealthy).

In [186]:
#Transform numeric value of 'status' column to categorical value for data_set
import pandas as pd
data_set['status'].replace(0, 'healthy', inplace = True)
data_set['status'].replace(1, 'unhealthy', inplace = True)
data_set['status']

name
phon_R01_S01_1    unhealthy
phon_R01_S01_2    unhealthy
phon_R01_S01_3    unhealthy
phon_R01_S01_4    unhealthy
phon_R01_S01_5    unhealthy
                    ...    
phon_R01_S50_2      healthy
phon_R01_S50_3      healthy
phon_R01_S50_4      healthy
phon_R01_S50_5      healthy
phon_R01_S50_6      healthy
Name: status, Length: 195, dtype: object

In [187]:
#Describe the data_set after transformation
data_set.describe()

MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
count   195.000000    195.000000    195.000000      195.000000   
mean    154.228641    197.104918    116.324631        0.006220   
std      41.390065     91.491548     43.521413        0.004848   
min      88.333000    102.145000     65.476000        0.001680   
25%     117.572000    134.862500     84.291000        0.003460   
50%     148.790000    175.829000    104.315000        0.004940   
75%     182.769000    224.205500    140.018500        0.007365   
max     260.105000    592.030000    239.170000        0.033160   

       MDVP:Jitter(Abs)    MDVP:RAP    MDVP:PPQ  Jitter:DDP  MDVP:Shimmer  \
count        195.000000  195.000000  195.000000  195.000000    195.000000   
mean           0.000044    0.003306    0.003446    0.009920      0.029709   
std            0.000035    0.002968    0.002759    0.008903      0.018857   
min            0.000007    0.000680    0.000920    0.002040      0.009540   
25%            0.000020    0.001660    0.001860    0.004985      0.016505   
50%            0.000030    0.002500    0.002690    0.007490      0.022970   
75%            0.000060    0.003835    0.003955    0.011505      0.037885   
max            0.000260    0.021440    0.019580    0.064330      0.119080   

       MDVP:Shimmer(dB)  ...    MDVP:APQ  Shimmer:DDA         NHR         HNR  \
count        195.000000  ...  195.000000   195.000000  195.000000  195.000000   
mean           0.282251  ...    0.024081     0.046993    0.024847   21.885974   
std            0.194877  ...    0.016947     0.030459    0.040418    4.425764   
min            0.085000  ...    0.007190     0.013640    0.000650    8.441000   
25%            0.148500  ...    0.013080     0.024735    0.005925   19.198000   
50%            0.221000  ...    0.018260     0.038360    0.011660   22.085000   
75%            0.350000  ...    0.029400     0.060795    0.025640   25.075500   
max            1.302000  ...    0.137780     0.169420    0.314820   33.047000   

             RPDE         DFA     spread1     spread2          D2         PPE  
count  195.000000  195.000000  195.000000  195.000000  195.000000  195.000000  
mean     0.498536    0.718099   -5.684397    0.226510    2.381826    0.206552  
std      0.103942    0.055336    1.090208    0.083406    0.382799    0.090119  
min      0.256570    0.574282   -7.964984    0.006274    1.423287    0.044539  
25%      0.421306    0.674758   -6.450096    0.174351    2.099125    0.137451  
50%      0.495954    0.722254   -5.720868    0.218885    2.361532    0.194052  
75%      0.587562    0.761881   -5.046192    0.279234    2.636456    0.252980  
max      0.685151    0.825288   -2.434031    0.450493    3.671155    0.527367  

[8 rows x 22 columns]

### Before normalization, I splited data into test set and training set. Then, divided them into numerical and categorical variable

In [188]:
#Split training set into training set and test set for data_set
from sklearn.model_selection import train_test_split
train_set, test_set = train_test_split(data_set, test_size = 0.2, random_state=42)

In [189]:
#Divided train_set and test_set into categorical and numerical set for data_set
train_set_num = train_set.drop('status', axis = 1)
train_set_cat = train_set["status"]
test_set_num = test_set.drop('status', axis = 1)
test_set_cat = test_set["status"]

### I am not sure that the result with normalization is more better than without normalization. I decided to do both and then compare them together to see how the results are.

In [190]:
#Using StandardScaler to normalize the numeric data set.
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
train_set_num_tr = scaler.fit_transform(train_set_num)
test_set_num_tr = scaler.fit_transform(test_set_num)

In [191]:
#Assign value into x_train, y_train, x_test, y_test (for variables without normalization)
x_train, y_train, x_test, y_test = train_set_num, train_set_cat, test_set_num, test_set_cat
x_train_nml, y_train_nml, x_test_nml, y_test_nml = train_set_num_tr, train_set_cat, test_set_num_tr, test_set_cat

## 4. Random Forests Classifier for data_set (without deleting columns)

### Default Hyperparameters

#### With normalization

In [192]:
#With normalization.
from sklearn.ensemble import RandomForestClassifier
rnd_clf_nml = RandomForestClassifier(random_state=42)
rnd_clf_nml.fit(x_train_nml, y_train_nml)

y_pred_rf_nml = rnd_clf_nml.predict(x_test_nml)

array(['unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'healthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'healthy', 'unhealthy', 'unhealthy', 'unhealthy'], dtype=object)

In [193]:
#Compute accuracy score for this prediction.
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test_nml, y_pred_rf_nml))

0.8974358974358975


In [194]:
#Compute F1-score
from sklearn.metrics import f1_score
rnd_nml_f1 = f1_score(y_test_nml, y_pred_rf_nml, average = 'macro')
rnd_nml_f1

0.803030303030303

#### Without normalization

In [223]:
#Without normalization.
from sklearn.ensemble import RandomForestClassifier
rnd_clf = RandomForestClassifier(random_state=42)
rnd_clf.fit(x_train, y_train)

y_pred_rf = rnd_clf.predict(x_test)

In [224]:
#Compute accuracy score for this prediction.
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred_rf))

0.9487179487179487


In [225]:
#Compute F1-score
from sklearn.metrics import f1_score
rnd_f1 = f1_score(y_test, y_pred_rf, average = 'macro')
rnd_f1

0.9015151515151515

### Change Hyperparameters

#### With normalzation

In [198]:
#With normalization.
from sklearn.ensemble import RandomForestClassifier
rnd_clf_nml1 = RandomForestClassifier(max_leaf_nodes=14, n_estimators=20, max_samples=27, bootstrap=True, n_jobs=-1, random_state=42)
rnd_clf_nml1.fit(x_train_nml, y_train_nml)

y_pred_rf_nml1 = rnd_clf_nml1.predict(x_test_nml)

array(['unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'healthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'healthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'healthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'healthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'healthy', 'unhealthy', 'unhealthy', 'unhealthy'], dtype=object)

In [199]:
#Compute accuracy score for this prediction.
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test_nml, y_pred_rf_nml1))

0.8205128205128205


In [200]:
#Compute F1-score
from sklearn.metrics import f1_score
rnd_nml1_f1 = f1_score(y_test_nml, y_pred_rf_nml1, average = 'macro')
rnd_nml1_f1

0.7367405978784957

#### Without normalization

In [201]:
#Without normalization.
from sklearn.ensemble import RandomForestClassifier
rnd_clf1 = RandomForestClassifier(max_leaf_nodes=14, n_estimators=20, max_samples=27, bootstrap=True, n_jobs=-1, random_state=42)
rnd_clf1.fit(x_train, y_train)

y_pred_rf1 = rnd_clf1.predict(x_test)

array(['unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'healthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'healthy', 'unhealthy', 'unhealthy', 'unhealthy'], dtype=object)

In [202]:
#Compute accuracy score for this prediction.
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred_rf1))

0.9230769230769231


In [203]:
#Compute F1-score
from sklearn.metrics import f1_score
rnd_f1 = f1_score(y_test, y_pred_rf1, average = 'macro')
rnd_f1

0.8615384615384616

## 5. Random Forests Classifier for data_set1 (deleting columns)

### Deleting columns

In [204]:
#Delete selected columns and assigned it into new variable data_set1
import pandas as pd
data_set1 = data_set.drop(['MDVP:Jitter(%)', 'MDVP:Jitter(Abs)', 'MDVP:RAP', 'MDVP:PPQ', 'Jitter:DDP', 'MDVP:Shimmer','Shimmer:APQ3', 
                          'Shimmer:APQ5','Shimmer:DDA', 'MDVP:APQ', 'NHR'], axis =1)
data_set1

MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Shimmer(dB)  \
name                                                                        
phon_R01_S01_1      119.992       157.302        74.997             0.426   
phon_R01_S01_2      122.400       148.650       113.819             0.626   
phon_R01_S01_3      116.682       131.111       111.555             0.482   
phon_R01_S01_4      116.676       137.871       111.366             0.517   
phon_R01_S01_5      116.014       141.781       110.655             0.584   
...                     ...           ...           ...               ...   
phon_R01_S50_2      174.188       230.978        94.261             0.405   
phon_R01_S50_3      209.516       253.017        89.488             0.263   
phon_R01_S50_4      174.688       240.005        74.287             0.256   
phon_R01_S50_5      198.764       396.961        74.904             0.241   
phon_R01_S50_6      214.289       260.277        77.973             0.190   

                   HNR     status      RPDE       DFA   spread1   spread2  \
name                                                                        
phon_R01_S01_1  21.033  unhealthy  0.414783  0.815285 -4.813031  0.266482   
phon_R01_S01_2  19.085  unhealthy  0.458359  0.819521 -4.075192  0.335590   
phon_R01_S01_3  20.651  unhealthy  0.429895  0.825288 -4.443179  0.311173   
phon_R01_S01_4  20.644  unhealthy  0.434969  0.819235 -4.117501  0.334147   
phon_R01_S01_5  19.649  unhealthy  0.417356  0.823484 -3.747787  0.234513   
...                ...        ...       ...       ...       ...       ...   
phon_R01_S50_2  19.517    healthy  0.448439  0.657899 -6.538586  0.121952   
phon_R01_S50_3  19.147    healthy  0.431674  0.683244 -6.195325  0.129303   
phon_R01_S50_4  17.883    healthy  0.407567  0.655683 -6.787197  0.158453   
phon_R01_S50_5  19.020    healthy  0.451221  0.643956 -6.744577  0.207454   
phon_R01_S50_6  21.209    healthy  0.462803  0.664357 -5.724056  0.190667   

                      D2       PPE  
name                                
phon_R01_S01_1  2.301442  0.284654  
phon_R01_S01_2  2.486855  0.368674  
phon_R01_S01_3  2.342259  0.332634  
phon_R01_S01_4  2.405554  0.368975  
phon_R01_S01_5  2.332180  0.410335  
...                  ...       ...  
phon_R01_S50_2  2.657476  0.133050  
phon_R01_S50_3  2.784312  0.168895  
phon_R01_S50_4  2.679772  0.131728  
phon_R01_S50_5  2.138608  0.123306  
phon_R01_S50_6  2.555477  0.148569  

[195 rows x 12 columns]

In [205]:
#Transform numeric value of 'status' column to categorical value for data_set1
data_set1['status'].replace(0, 'healthy', inplace = True)
data_set1['status'].replace(1, 'unhealthy', inplace = True)
data_set1['status']

name
phon_R01_S01_1    unhealthy
phon_R01_S01_2    unhealthy
phon_R01_S01_3    unhealthy
phon_R01_S01_4    unhealthy
phon_R01_S01_5    unhealthy
                    ...    
phon_R01_S50_2      healthy
phon_R01_S50_3      healthy
phon_R01_S50_4      healthy
phon_R01_S50_5      healthy
phon_R01_S50_6      healthy
Name: status, Length: 195, dtype: object

In [206]:
#Describe the data_set1 after transformation
data_set1.describe()

MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Shimmer(dB)         HNR  \
count   195.000000    195.000000    195.000000        195.000000  195.000000   
mean    154.228641    197.104918    116.324631          0.282251   21.885974   
std      41.390065     91.491548     43.521413          0.194877    4.425764   
min      88.333000    102.145000     65.476000          0.085000    8.441000   
25%     117.572000    134.862500     84.291000          0.148500   19.198000   
50%     148.790000    175.829000    104.315000          0.221000   22.085000   
75%     182.769000    224.205500    140.018500          0.350000   25.075500   
max     260.105000    592.030000    239.170000          1.302000   33.047000   

             RPDE         DFA     spread1     spread2          D2         PPE  
count  195.000000  195.000000  195.000000  195.000000  195.000000  195.000000  
mean     0.498536    0.718099   -5.684397    0.226510    2.381826    0.206552  
std      0.103942    0.055336    1.090208    0.083406    0.382799    0.090119  
min      0.256570    0.574282   -7.964984    0.006274    1.423287    0.044539  
25%      0.421306    0.674758   -6.450096    0.174351    2.099125    0.137451  
50%      0.495954    0.722254   -5.720868    0.218885    2.361532    0.194052  
75%      0.587562    0.761881   -5.046192    0.279234    2.636456    0.252980  
max      0.685151    0.825288   -2.434031    0.450493    3.671155    0.527367

### Before normalization, I splited data into test set and training set. Then, divided them into numerical and categorical variable

In [207]:
#Split training set into training set and test set for data_set1
from sklearn.model_selection import train_test_split
train_set1, test_set1 = train_test_split(data_set1, test_size = 0.2, random_state=42)

In [208]:
#Divided train_set and test_set into categorical and numerical set for data_set1
train_set_num1 = train_set1.drop('status', axis = 1)
train_set_cat1 = train_set1["status"]
test_set_num1 = test_set1.drop('status', axis = 1)
test_set_cat1 = test_set1["status"]

### I am not sure that the result with normalization is more better than without normalization. I decided to do both and then compare them together to see how the results are.

In [209]:
#Using StandardScaler to normalize the numeric data set.
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
train_set_num_tr1 = scaler.fit_transform(train_set_num1)
test_set_num_tr1 = scaler.fit_transform(test_set_num1)

In [210]:
#Assign value into x_train1, y_train1, x_test1, y_test1 (for variables without normalization)
x_train1, y_train1, x_test1, y_test1 = train_set_num1, train_set_cat1, test_set_num1, test_set_cat1
x_train_nml1, y_train_nml1, x_test_nml1, y_test_nml1 = train_set_num_tr1, train_set_cat1, test_set_num_tr1, test_set_cat1

### Default Hyperparameters

#### With normalization

In [211]:
#With normalization.
from sklearn.ensemble import RandomForestClassifier
rnd_clf_nml = RandomForestClassifier(random_state=42)
rnd_clf_nml.fit(x_train_nml1, y_train_nml1)

y_pred_rf_nml = rnd_clf_nml.predict(x_test_nml1)

array(['unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'healthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'healthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'healthy', 'unhealthy', 'unhealthy', 'unhealthy'], dtype=object)

In [212]:
#Compute accuracy score for this prediction.
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test_nml1, y_pred_rf_nml))

0.8717948717948718


In [213]:
#Compute F1-score
from sklearn.metrics import f1_score
rnd_nml_f1 = f1_score(y_test_nml1, y_pred_rf_nml, average = 'macro')
rnd_nml_f1

0.7692307692307692

#### Without normalization

In [226]:
#Without normalization.
from sklearn.ensemble import RandomForestClassifier
rnd_clf = RandomForestClassifier(random_state=42)
rnd_clf.fit(x_train1, y_train1)

y_pred_rf = rnd_clf.predict(x_test1)

In [227]:
#Compute accuracy score for this prediction.
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test1, y_pred_rf))

0.9487179487179487


In [228]:
#Compute F1-score
from sklearn.metrics import f1_score
rnd_f1 = f1_score(y_test1, y_pred_rf, average = 'macro')
rnd_f1

0.9015151515151515

### Change Hyperparameters

#### With normalization

In [217]:
#With normalization.
from sklearn.ensemble import RandomForestClassifier
rnd_clf_nml1 = RandomForestClassifier(max_leaf_nodes=14, n_estimators=20, max_samples=27, bootstrap=True, n_jobs=-1, random_state=42)
rnd_clf_nml1.fit(x_train_nml1, y_train_nml1)

y_pred_rf_nml1 = rnd_clf_nml1.predict(x_test_nml1)

array(['unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'healthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'healthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'healthy', 'unhealthy', 'unhealthy', 'unhealthy'], dtype=object)

In [218]:
#Compute accuracy score for this prediction.
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test_nml1, y_pred_rf_nml1))

0.8461538461538461


In [219]:
#Compute F1-score
from sklearn.metrics import f1_score
rnd_nml1_f1 = f1_score(y_test_nml1, y_pred_rf_nml1, average = 'macro')

0.7388392857142857

#### Without normalization

In [220]:
#Without normalization.
from sklearn.ensemble import RandomForestClassifier
rnd_clf1 = RandomForestClassifier(max_leaf_nodes=14, n_estimators=20, max_samples=27, bootstrap=True, n_jobs=-1, random_state=42)
rnd_clf1.fit(x_train1, y_train1)

y_pred_rf1 = rnd_clf1.predict(x_test1)

array(['unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'healthy', 'unhealthy',
       'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy', 'unhealthy',
       'healthy', 'unhealthy', 'unhealthy', 'unhealthy'], dtype=object)

In [221]:
#Compute accuracy score for this prediction.
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test1, y_pred_rf1))

0.8974358974358975


In [222]:
#Compute F1-score
from sklearn.metrics import f1_score
rnd_f1 = f1_score(y_test1, y_pred_rf1, average = 'macro')
rnd_f1

0.7705882352941176

## 6. Conclusion

| Random Forests           |                          |                       | Accuracy score | F1-score |
|--------------------------|:------------------------:|-----------------------|:--------------:|:--------:|
| Without deleting columns |  Default Hyperparameters | With normalization    |     89.74%     |  80.30%  |
|                          |                          | Without normalization |     94.87%     |  90.15%  |
|                          | Changing Hyperparameters | With normalization    |     82.05%     |  73.67%  |
|                          |                          | Without normalization |     92.31%     |  86.15%  |
| With deleting columns    |  Default Hyperparameters | With normalization    |     87.18%     |  76.92%  |
|                          |                          | Without normalization |     94.87%     |  90.15%  |
|                          | Changing Hyperparameters |   With normalization  |     84.62%     |  73.88%  |
|                          |                          | Without normalization |     89.74%     |  77.06%  |

_ At first, let compare the Accuracy score and F1-score between keeping the columns and deleting columns, we can easily see the scores that tended to decrease if you used the data set with deleting columns step. Base on this, I can give out conclusion that we should not delete columns in this dataset where we could get the better scores instead of receiving lower scores.

_ Secondly, there are some differences in both scores between default hyperparameters and changing hyperparameters. The scores tended to decline from Default Hyperparameters to Changing Hyperparameters. For example, the scores decreased from 89.74%(default) to 82.05%(changing) for the dataset without deliting columns or the scores decreased from 87.18%(default) to 84.62%(changing).

_ Thirdly, it seems that when doing Random Forests Classification, we do not need to do normalization step because I can easily at the table that the Accuracy and F1 scores without normalization are always higher than scores with normalization in both data set of keeping columns and deleting columns. For example, 89.74%(with normalization) increased to 94.84%(no normalization) in the Default Hyperparameters section. You can easily have a look at the table above to see how the scores changed in each test.
_ Fourly, I recognized that there is a special case in this dataset. Even though I deleted or not deleted the columns and kept the default hyperparameters for both case without normalization, the scores of these cases are still the same 94.87%(Accuracy) 90.15%(F1-score). I thought that the aforementioned columns I deleted might not affect the scores compared to the scores of the data set without deleting the columns. But others situations they might be affect the scores. 

=> So to not waste the time, I decided to do with the data set with default hyperparameters and without deleting columns and without normalization to get the best scores for prediction. I chose the way in row 2 in the table that I could get 94.87% of Accuracy score and 90.15% of F1-score. Both scores are above 90%, which means that these scores are above the perfect rate for correct prediction.